snuba/consumers/strict_consumer.py

```diff
@@ -0,0 +1,141 @@
+            'default.topic.config': {
+                'auto.offset.reset': auto_offset_reset,
+            },
+        }
+
+        self.__consumer = self.create_consumer(consumer_config)
+
+        def on_partitions_assigned(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            logger.info("New partitions assigned: %r", partitions)
+            self.__on_partitions_assigned(consumer, partitions)
+
+        def on_partitions_revoked(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            logger.info("Partitions revoked: %r", partitions)
+            self.__on_partitions_revoked(consumer, partitions)
+
+        self.__consumer.subscribe(
+            [topic],
+            on_assign=on_partitions_assigned,
+            on_revoke=on_partitions_revoked,
+        )
+
+    def create_consumer(self, config) -> Consumer:
+        return Consumer(config)
+
+    def run(self) -> None:
+        watermarks = {}
+
+        message = self.__consumer.poll(timeout=self._assign_timeout)
+        if not message:
+            # Since we enabled enable.partition.eof, no message means I could not
+            # consume and it is different from the case I managed to consume and
+            # I reached the end of the partition. Thus we fail.
+            raise NoPartitionAssigned("No partition was assigned within timeout")
+
+        while not self.__shutdown:
+            error = message.error()
+            if error:
+                if error.code() == KafkaError._PARTITION_EOF:
+                    logger.debug("End of topic reached")
+                    self.__consumer.close()
+                    return
+                else:
+                    raise Exception(message.error())
+
+            commit_decision = self.__on_message(message)
+            if commit_decision == CommitDecision.COMMIT_THIS:
+                self.__consumer.commit()
+            elif commit_decision == CommitDecision.COMMIT_PREV:
+                prev_watermark = watermarks.get(
+                    (message.partition(), message.topic()),
+                )
+                if prev_watermark is not None:
+                    commit_pos = TopicPartition(
+                        partition=message.partition(),
+                        topic=message.topic(),
+                        offset=prev_watermark,
+                    )
+                    self.__consumer.commit(commit_pos)
+                else:
+                    logger.debug(
+                        "No previous message to commit on partition %s on topic %s",
+                        message.partition(),
+                        message.topic(),
+                    )
+
+            watermarks[(message.partition(), message.topic())] = message.offset()
+            message = self.__consumer.poll(timeout=1.0)
+
+        self.__consumer.close()
+
+    def force_commit(self, partition: TopicPartition) -> None:
```

Comment:
What's this used for? It looks unused and seems like it could be dangerous if used incorrectly, so I'd err on the side of not including it unless there's a specific use coming up.

Reply:
Earlier implementation, we can remove it for now. Though at a point above in the stack something like this will be needed when we want to commit after having consumed the snapshot. But we can avoid specifying the offset.